src/main/java/ch/njol/skript/expressions/ExprLore.java

```diff
@@ -1,236 +1,181 @@
-
-		// air is just nothing, it can't have a lore
-		if (i == null || i instanceof ItemStack && ((ItemStack) i).getType() == Material.AIR)
-			return;
-
-		ItemMeta meta = i instanceof ItemStack ? ((ItemStack) i).getItemMeta() : (ItemMeta) ((ItemType) i).getItemMeta();
-		if (meta == null)
-			meta = Bukkit.getItemFactory().getItemMeta(Material.STONE);
-
-		Number lineNumber = this.lineNumber != null ? this.lineNumber.getSingle(e) : null;
-		List<String> lore = meta.hasLore() ? new ArrayList<>(meta.getLore()) : new ArrayList<>();
-
-		if (lineNumber == null) {
-			// if the condition below is true, the pattern with the line %number% expression was used,
-			// but the line number turned out to be null at runtime, meaning we should ignore it
-			if (this.lineNumber != null) {
-				return;
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		if (this.item.getTime() != EventValues.TIME_NOW && !ChangerUtils.acceptsChange(item, ChangeMode.SET, ItemStack.class, ItemType.class)) {
+			// Allows past/future expressions only if they accept the set change mode
+			return null;
+		}
+		boolean acceptsMany = this.lineNumber == null;
+		return switch (mode) {
+			case SET -> CollectionUtils.array(acceptsMany ? String[].class : String.class);
+			case DELETE -> CollectionUtils.array();
+			case ADD, REMOVE, REMOVE_ALL -> {
+				if (!acceptsMany) {
+					Skript.error("You cannot add/remove the lore from a single line, you can however replace/concat lore.");
+					yield null;
+				}
+				yield CollectionUtils.array(String[].class);
 			}
+			default -> null;
+		};
+	}
 
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		Object item = this.item.getSingle(event);
+		if (!validateItem(item))
+			return; // Validates to ensure it is a valid item and has item meta.
+		ItemStack modifiedItem = ItemUtils.asItemStack(item);
+		assert modifiedItem != null; // validateItem has already run a check against this
+		ItemMeta itemMeta = modifiedItem.getItemMeta();
+		//noinspection deprecation
+		List<String> modifiedLore = itemMeta.hasLore() ? itemMeta.getLore() : new ArrayList<>();
+		assert modifiedLore != null; // lore can never be null here, if it's unset we create an empty list
+
+		if (this.lineNumber == null) {
 			switch (mode) {
-				case SET:
-					assert stringDelta != null;
-					List<String> newLore = new ArrayList<>();
-					for (String line : stringDelta) {
-						if (line.contains("\n")) {
-							Collections.addAll(newLore, line.split("\n"));
-							continue;
-						}
-						newLore.add(line);
-					}
-					lore = newLore;
-					break;
-				case ADD:
-					assert stringDelta != null;
-					List<String> addLore = new ArrayList<>();
-					for (String line : stringDelta) {
-						if (line.contains("\n")) {
-							Collections.addAll(addLore, line.split("\n"));
-							continue;
-						}
-						addLore.add(line);
+				case SET, DELETE -> modifiedLore = (delta != null) ? List.of((String[]) delta) : null;
```

Comment:
Casting delta is not guaranteed to be safe unfortunately, so you may need a different method of accomplishing this

Reply:
Do you have any suggestions for this? SInce delta can only ever be string unless someone does an unsafe call manually through `new ExprClass().change()` or `Changer#change()`. The only other way I could think of going through is
```java
List<String> providedStrings = new ArrayList<>();
if (delta != null) {
  for (Object object : delta) {
    if (!(object instanceof String stringDelta)
      continue;
    providedStrings.add(stringDelta);
  }
} else {
  providedStrings = null;
}
```